scm-core/src/main/java/sonia/scm/repository/work/SimpleCachingWorkingCopyPool.java

```diff
@@ -0,0 +1,125 @@
+ * Copyright (c) 2020-present Cloudogu GmbH and Contributors
+ *
+ * Permission is hereby granted, free of charge, to any person obtaining a copy
+ * of this software and associated documentation files (the "Software"), to deal
+ * in the Software without restriction, including without limitation the rights
+ * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+ * copies of the Software, and to permit persons to whom the Software is
+ * furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in all
+ * copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+ * SOFTWARE.
+ */
+
+package sonia.scm.repository.work;
+
+import com.google.common.base.Stopwatch;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import sonia.scm.util.IOUtil;
+
+import javax.inject.Inject;
+import java.io.File;
+import java.util.Map;
+import java.util.concurrent.ConcurrentHashMap;
+
+/**
+ * This class is a simple implementation of the {@link WorkingCopyPool} to demonstrate,
+ * how caching can work in the simplest way. For the first time a {@link WorkingCopy} is
+ * requested for a repository with {@link #getWorkingCopy(SimpleWorkingCopyFactory.WorkingCopyContext)},
+ * this implementation fetches a new directory from the {@link WorkdirProvider}.
+ * On {@link #contextClosed(SimpleWorkingCopyFactory.WorkingCopyContext, File)},
+ * the directory is not deleted, buy put into a map with the repository id as key.
+ * When a working copy is requested with {@link #getWorkingCopy(SimpleWorkingCopyFactory.WorkingCopyContext)}
+ * for a repository with such an existing directory, it is taken from the map, reclaimed and
+ * returned as {@link WorkingCopy}.
+ * If for one repository a working copy is requested, while another is in use already,
+ * a second directory is requested from the {@link WorkdirProvider} for the second one.
+ * If a context is closed with {@link #contextClosed(SimpleWorkingCopyFactory.WorkingCopyContext, File)}
+ * and there already is a directory stored in the map for the repository,
+ * the directory from the closed context simply is deleted.
+ * <br>
+ * In general, this implementation should speed up things a bit, but one has to take into
+ * account, that there is no monitoring of diskspace. So you have to make sure, that
+ * there is enough space for a clone of each repository in the working dir.
+ * <br>
+ * Possible enhancements:
+ * <ul>
+ *   <li>Monitoring of times</li>
```

Comment:
What do you mean by "Monitoring of times"? Do you mean more than the Stopwatch logging, which is already implemented?

Reply:
Yeah, maybe you can include the time needed to clone a repository when you have to dismiss cached working directories, or you can even use caches for single branches when switchen between them takes too long.